// librarian/lib/Librarian.js
// Librarian

var path        = require('path')
var fs          = require('fs')
var temp        = require('temp')

var Joose       = require('../../../joose-all')

var File        = require('./File')



module.exports = Joose.Class({
    
    does    : require('./ModuleResolver')(Joose),
    
    has     : {
        coreModules     : {
            init        : {
                'fs'    : true,
                'path'  : true,
                'event' : true
            }
        },
        
        root            : { required : true },
        
        files           : Joose.I.Object,
        
        tempRoot        : {
            
            lazy : function () {
            
            }
        }
    },
    
    
    override : {
        
        resolve : function (require, baseDir) {
            if (baseDir instanceof File) baseDir = path.dirname(baseDir.name)
            
            return this.resolve(require, baseDir)
        }
    },
    
    
    methods : {
        
        initialize : function () {
            this.root       = path.resolve(this.root)
        },
        
        
        getFileAsJSON : function (fileName) {
            return JSON.parse(this.getContentOf(fileName))
        },
        
        
        getContentOf : function (fileName) {
            return fs.readFileSync(path.join(this.root, fileName), 'utf8')
        },
        
        
//        normalizeFileName : function (fileName) {
//            return path.relative(this.root, path.resolve(fileName))
//        },
        
        getCoreModule : function (coreModule) {
            throw new Error('not supported yet')
        },
        
        
        fileExists : function (fileName) {
            try {
                var stat    = fs.statSync(path.join(this.root, fileName))
                
                return stat.isFile()
                
            } catch (e) {
                return false
            }
        },
        
        
        getFile : function (fileName) {
            var files   = this.files
            
            if (files[ fileName ]) return fileName
        
            return files[ fileName ] = new File({
                name        : fileName,
                librarian   : this
            })
        },
        
        
//        resolveAsDir : function () {
//        },
//        
//        
//        resolveAsFile : function () {
//        },
//        
//        
//        resolveAsModule : function () {
//        },
//        
//        
//        resolve : function (require, baseDir) {
//            if (baseDir instanceof File) baseDir = path.dirname(baseDir.name)
//            
//            return this.resolve(require, baseDir)
//        },
        
        wrap : function (file) {
            return '<wrap>' + me.getContentOf(file) + '</wrap>'
        },
        
        
        writeBundleTo : function (stream, requires) {
            var me      = this
            
            var files   = Joose.A.map(requires, function (require) {
                
                return me.getFile(me.resolve(require, '.'))
            })
            
            var allFiles    = {}
            
            Joose.A.each(files, function (file) {
                
                Joose.O.each(file.getInDepthDependencies(), function (file, name) {
                    if (allFiles[ name ] && allFiles[ name ] != file) throw new Error('Inconsistency')
                    
                    allFiles[ name ] = file
                })
            })
            
            Joose.O.each(allFiles, function (file) {
                stream.write(me.wrap(file) + ';\n')
            })
        }
    }
})
